### PR TITLE
Vulkan: Fix incorrect access to the buffers on Android

### LINKED ISSUE #75599: Mobile Renderer makes Meshes unreadable/uneditable on Android
### Godot version

4.0.1-stable

### System information

Android 12, Mobile Renderer

### Issue description

When retrieving Mesh data from a MeshInstance3D like so...
```GDScript
var mesh = self.get_mesh()
var mesh_array = mesh.surface_get_arrays(0)
var vertex_array = mesh_array[Mesh.ARRAY_VERTEX]
var index_array = mesh_array[Mesh.ARRAY_INDEX]
```
...the index_array contains vertex indices that are not within the vertex_array.

This issue does not occur in the Linux editor, but does occur on my Android build:
Android 12
arm64v8
sdk version 32.0.0
ndk version 23.2.8568313

### Steps to reproduce

Download the minimal reproduction project, which contains an example ArrayMesh and GDScript script which has the code shown above. Build to Android using the 4.0.1 build template and test.

### Minimal reproduction project

[Array_Mesh_Test.zip](https://github.com/godotengine/godot/files/11133179/Array_Mesh_Test.zip)

### LINKED ISSUE #78715: 4.1 Crash when using SoftBody on mobile
### Godot version

4.1 beta 3, 4.1 beta 1

### System information

Windows 10 - 4.1 beta 3 - Android 12

### Issue description

Crash when using SoftBody on android mobile.

### Steps to reproduce

- Simple scene setup using a plane softbody, camera.
- Export to android mobile.
- Immediate crash when open the app.

### Minimal reproduction project

[SoftbodyMobileTest.zip](https://github.com/godotengine/godot/files/11871734/SoftbodyMobileTest.zip)

### LINKED ISSUE #80371: OBJ meshes get corrupted on reimport on Android editor
### Godot version

4.1.stable

### System information

Android13

### Issue description

I'm trying to import the obj model file on Android but there seems to be some compatibility issues? You can do two tests on both the computer and the mobile side to make sure what I'm saying is true.
Screenshot_2023-08-07-19-44-12-648_org godotengine editor v4

### Steps to reproduce

Create an obj file and import it into the Godot engine on Android.

### Minimal reproduction project
[test.zip](https://github.com/godotengine/godot/files/12279298/test.zip)

### LINKED ISSUE #84355: Surface Tool doesn't generate mesh on mobile renderer with Android 10
### Godot version

v4.2.beta4.official [93cdacbb0], 4.1.2.stable

### System information

Godot v4.2.beta4 - Android 10 - Vulkan (Mobile)

### Issue description

This happens only when trying to run it on Android 10 using the mobile renderer.

It fails on Samsung Galaxy S9+ (Android 10).
It works on Nokia 5.4 (Android 12).

The surface tool doesn't seem generate mesh at least not visible one. It also fails to generate collision mesh as a character will fall through it.

I included an example code. 

### Steps to reproduce

Export the attached reproduction project (remember to set the renderer to mobile) and install it on an Android phone. The phone I tested is Samsung Galaxy S9+ with Android 10.

### Minimal reproduction project

[SurfaceToolBug.zip](https://github.com/godotengine/godot/files/13237789/SurfaceToolBug.zip)

### PR DESCRIPTION
**Pull request by Alex2782**
Fixes #78715
Fixes #75599
Fixes #80371
Fixes #84355

**Error description:** https://github.com/godotengine/godot/issues/78715#issuecomment-1793603910 / https://github.com/godotengine/godot/issues/78715#issuecomment-1807297175

- **Test project** [test.buffer.zip](https://github.com/godotengine/godot/files/13349980/test.buffer.zip) for **78715**, **75599** and **84355**
- 41 MB download (drive.google.com) [android_editor-debug.apk](https://drive.google.com/file/d/1nU4DTZTTaukpxyQ_JssT8ynenWMsD359/view) for **80371**

-----------------------

<details><summary>old tests</summary>

**Godot 4.2 beta5** **78715** and **75599**
"Mesh Array_Index" is not read correctly (it sometimes contains 'attrib_data' or 'vertex_data'). The app crashes when "soft_body_3d" is inserted.

https://github.com/godotengine/godot/assets/41921395/36fb7eb8-b281-49d5-826e-802992d1c79a

**this PR** **78715** and **75599**
"Mesh Array_Index" is always read correctly and there are no crashes when soft_body_3d is inserted.

https://github.com/godotengine/godot/assets/41921395/ec864576-d0ea-4afa-a490-726b1b21416b

-----------------------
**80371** Godot 4.1.3

https://github.com/godotengine/godot/assets/41921395/bc266921-336c-406c-b78e-a7567a6e66be

**80371** Godot 4.2 (beta.custom, this PR)

https://github.com/godotengine/godot/assets/41921395/07f70a57-92f6-4183-9277-69519012c8ba

</details>

### CODE REVIEW

File: drivers/vulkan/rendering_device_vulkan.h [hunk 1]
```diff
@@ -43,7 +43,7 @@
 #define _DEBUG
 #endif
 #endif
-#include "vk_mem_alloc.h"
+#include "thirdparty/vulkan/vk_mem_alloc.h"
```
Comment (akien-mga) [inline drivers/vulkan/rendering_device_vulkan.h:46]: Does this change fix something specific?

`driver/vulkan/SCsub` already adds `thirdparty/vulkan` in the include path so it doesn't need to be specified explicitly (same as `volk.h` and `vulkan/vulkan.h` below). This could use `<vk_mem_alloc.h>` instead of `"vk_mem_alloc.h"` to be consistent though.

File: drivers/vulkan/rendering_device_vulkan.cpp [hunk 2]
```diff
@@ -6082,7 +6082,13 @@ Vector<uint8_t> RenderingDeviceVulkan::buffer_get_data(RID p_buffer, uint32_t p_
 	}
 
 	Buffer tmp_buffer;
+
+#ifdef __ANDROID__
+	_buffer_allocate(&tmp_buffer, p_size, VK_BUFFER_USAGE_TRANSFER_DST_BIT, VMA_MEMORY_USAGE_CPU_ONLY, VMA_ALLOCATION_CREATE_HOST_ACCESS_RANDOM_BIT);
```
Comment (clayjohn) [inline drivers/vulkan/rendering_device_vulkan.cpp:6087]: Its nice to see this issue getting some attention! And great that you have identified a solution. However, I am a bit concerned about using the `VMA_MEMORY_USAGE_CPU_ONLY` bit as it is [deprecated](https://gpuopen-librariesandsdks.github.io/VulkanMemoryAllocator/html/group__group__alloc.html#ggaa5846affa1e9da3800e3e78fae2305cca40bdf4cddeffeb12f43d45ca1286e0a5)

I have a feeling that ``VMA_MEMORY_USAGE_AUTO_PREFER_HOST`` is the correct option and we need to add another flag to the last parameter. The [docs](https://gpuopen-librariesandsdks.github.io/VulkanMemoryAllocator/html/usage_patterns.html) seem to indicate ``VMA_ALLOCATION_CREATE_HOST_ACCESS_RANDOM_BIT | VMA_ALLOCATION_CREATE_MAPPED_BIT``

Comment (clayjohn) [reply drivers/vulkan/rendering_device_vulkan.cpp:6087]: Okay, there could be something deeper going on here.

`VMA_MEMORY_USAGE_CPU_ONLY` guarantees that `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT` is set. ``VK_MEMORY_PROPERTY_HOST_COHERENT_BIT`` allows us to get away without flushing or invalidating mapped memory regions using [vkFlushMappedMemoryRanges](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkFlushMappedMemoryRanges.html) and [vkInvalidateMappedMemoryRanges](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkInvalidateMappedMemoryRanges.html) respectively.

After allocating the memory, we do a GPU flush, but we don't call either of those functions. I wonder if the core of the issue is that we aren't properly making the memory available for reading. `VMA_MEMORY_USAGE_CPU_ONLY` works around the issue by using `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT`, but since its deprecated, we should probably implement an actual solution.

Comment (Alex2782) [reply drivers/vulkan/rendering_device_vulkan.cpp:6087]: https://github.com/godotengine/godot/issues/78715#issuecomment-1807297175


Another solution would be to assign `requiredFlags`. If it is to be parameterized, then I think significantly more adjustments are required.

https://github.com/godotengine/godot/blob/59457685c18e2d729eea50c751c11f049a7186f0/drivers/vulkan/rendering_device_vulkan.cpp#L1363

`allocInfo.requiredFlags = VK_MEMORY_PROPERTY_HOST_COHERENT_BIT`

Or are there other solutions?

`VMA_MEMORY_USAGE_CPU_ONLY` 
> Obsolete, preserved for **backward compatibility.** Guarantees VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT and VK_MEMORY_PROPERTY_HOST_COHERENT_BIT.

`Vulkan API 1.2.231` on MacOS M1
`Vulkan API 1.1.128` on Android. Maybe it is generally necessary with `version 1.1`, does anyone have an old graphics card for example?

Comment (clayjohn) [reply drivers/vulkan/rendering_device_vulkan.cpp:6087]: Okay, another update that explains things a bit more. It seems that on Desktop devices HOST_VISIBLE is pretty much always paired with HOST_COHERENT (whether you request it or not). While on Android, you can have HOST_VISIBLE without HOST_COHERENT (so presumably, you need to ask for it)

https://vulkan.gpuinfo.org/listmemory.php?platform=android

So we have two possible alternative solutions:
1. Ask for ``VK_MEMORY_PROPERTY_HOST_COHERENT_BIT`` on Android devices (by modifying ``_buffer_allocate()``
2. Actually use [vkFlushMappedMemoryRanges](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkFlushMappedMemoryRanges.html) and [vkInvalidateMappedMemoryRanges](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkInvalidateMappedMemoryRanges.html) instead of relying on ``VK_MEMORY_PROPERTY_HOST_COHERENT_BIT``

Comment (clayjohn) [reply drivers/vulkan/rendering_device_vulkan.cpp:6087]: > Or are there other solutions?

The other solution is to use flush/invalidate I guess. I'm not sure if we are guaranteed to have HOST_VISIBLE memory available that also has ``VK_MEMORY_PROPERTY_HOST_COHERENT_BIT``. If we are, then we can just make ``VK_MEMORY_PROPERTY_HOST_COHERENT_BIT`` required whenever we read back from the GPU, if not then the solution will be more complex

Comment (Alex2782) [reply drivers/vulkan/rendering_device_vulkan.cpp:6087]: > Edit: Sorry, I wrote all three comments without refreshing my page,

I also missed some comments, Github often updates automatically, not today. In the next few days I'll try the first solution. 

Perhaps this function should also be adapted.
https://github.com/godotengine/godot/blob/59457685c18e2d729eea50c751c11f049a7186f0/drivers/vulkan/rendering_device_vulkan.cpp#L1399

```cpp
	allocInfo.flags = VMA_ALLOCATION_CREATE_HOST_ACCESS_SEQUENTIAL_WRITE_BIT;
	allocInfo.usage = VMA_MEMORY_USAGE_AUTO_PREFER_HOST;
```

Comment (Alex2782) [reply drivers/vulkan/rendering_device_vulkan.cpp:6087]: I'll try this first, the VMA description is easier for me to understand

--------------
[vmaInvalidateAllocation](https://gpuopen-librariesandsdks.github.io/VulkanMemoryAllocator/html/group__group__alloc.html#gaaa8412919139ef413a4215ac6a290fae) -> Calls `vkInvalidateMappedMemoryRanges` 
_It needs to be called **before reading** from a mapped memory for memory types that are not HOST_COHERENT. Map operation doesn't do that automatically._

[vmaFlushAllocation](https://gpuopen-librariesandsdks.github.io/VulkanMemoryAllocator/html/group__group__alloc.html#ga30c37c1eec6025f397be41644f48490f)  -> Calls `vkFlushMappedMemoryRanges`
_It needs to be called **after writing** to a mapped memory for memory types that are not HOST_COHERENT. Unmap operation doesn't do that automatically._

**Comment by Alex2782** [comment]
The functions `vmaInvalidateAllocation` / `vmaFlushAllocation` require too much performance. The FPS goes from 60 to 40 because 'issue **75599**' queries the data via `_process`. With `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT` this was no problem.

```c++
	void *buffer_mem;
	VkResult vkerr = vmaMapMemory(allocator, tmp_buffer.allocation, &buffer_mem);
	ERR_FAIL_COND_V_MSG(vkerr, Vector<uint8_t>(), "vmaMapMemory failed with error " + itos(vkerr) + ".");

#ifdef __ANDROID__
	vmaInvalidateAllocation(allocator, tmp_buffer.allocation, 0, VK_WHOLE_SIZE);
#endif

	Vector<uint8_t> buffer_data;
	{
		buffer_data.resize(p_size);
		uint8_t *w = buffer_data.ptrw();
		memcpy(w, buffer_mem, p_size);
	}

#ifdef __ANDROID__
	vmaFlushAllocation(allocator, tmp_buffer.allocation, 0, VK_WHOLE_SIZE);
#endif

	vmaUnmapMemory(allocator, tmp_buffer.allocation);

	_buffer_free(&tmp_buffer);
```


https://github.com/godotengine/godot/assets/41921395/24c7600f-faa6-4bb3-b965-6a7740a06d40

**Comment by clayjohn** [comment]
> The functions `vmaInvalidateAllocation` / `vmaFlushAllocation` require too much performance. The FPS goes from 60 to 40 because 'issue **75599**' queries the data via `_process`. With `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT` this was no problem.
> 
> ```c++
> 	void *buffer_mem;
> 	VkResult vkerr = vmaMapMemory(allocator, tmp_buffer.allocation, &buffer_mem);
> 	ERR_FAIL_COND_V_MSG(vkerr, Vector<uint8_t>(), "vmaMapMemory failed with error " + itos(vkerr) + ".");
> 
> #ifdef __ANDROID__
> 	vmaInvalidateAllocation(allocator, tmp_buffer.allocation, 0, VK_WHOLE_SIZE);
> #endif
> 
> 	Vector<uint8_t> buffer_data;
> 	{
> 		buffer_data.resize(p_size);
> 		uint8_t *w = buffer_data.ptrw();
> 		memcpy(w, buffer_mem, p_size);
> 	}
> 
> #ifdef __ANDROID__
> 	vmaFlushAllocation(allocator, tmp_buffer.allocation, 0, VK_WHOLE_SIZE);
> #endif
> 
> 	vmaUnmapMemory(allocator, tmp_buffer.allocation);
> 
> 	_buffer_free(&tmp_buffer);
> ```
> 
>  Screen_recording_20231113_231241.mp4

Then let's just request COHERENT memory. I would add an optional parameter to _buffer_allocate() called `p_coherent` (default to false). When true it should add the coherent flag to the preferred flags

**Comment by Alex2782** [comment]
**now**: `requiredFlags = VK_MEMORY_PROPERTY_HOST_COHERENT_BIT | VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT;`
**like:** [google-hellovk](https://github.com/android/ndk-samples/blob/87bf557531f3d3d50fbba6d64ac4a601a865788a/hello-vulkan/app/src/main/cpp/hellovk.h#L363) and [VMA_MEMORY_USAGE_CPU_ONLY](https://github.com/godotengine/godot/blob/64150060f89677eaf11229813ae6c5cf8a873802/thirdparty/vulkan/vk_mem_alloc.h#L3671C36-L3671C36)

**Tests:**

- without FPS limit between 63 and 69
- without `set_text(debug)` _(for issue 75599, RichText, gdscript)_ 20 FPS more
- without issue 75599 `self.get_mesh()` and `mesh_data.surface_get_arrays(0)` up to 120 fps _(system limit)_

**Review by Alex2782 — COMMENTED** [review]
**Higher performance(+25 fps)**: `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT | VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT`

File: drivers/vulkan/rendering_device_vulkan.cpp [hunk 3]
```diff
@@ -6082,7 +6085,12 @@ Vector<uint8_t> RenderingDeviceVulkan::buffer_get_data(RID p_buffer, uint32_t p_
 	}
 
 	Buffer tmp_buffer;
-	_buffer_allocate(&tmp_buffer, p_size, VK_BUFFER_USAGE_TRANSFER_DST_BIT, VMA_MEMORY_USAGE_AUTO_PREFER_HOST, VMA_ALLOCATION_CREATE_HOST_ACCESS_RANDOM_BIT);
+#ifdef __ANDROID__
+	bool p_coherent = true;
```
Comment (lmurray) [inline drivers/vulkan/rendering_device_vulkan.cpp:6089]: It seems odd to me that this is only being used on a single platform when the API is supposed to be platform-independent. If this is needed for a particular hardware configuration shouldn't we be testing for that configuration instead of just testing for the platform? If there is no way to test for the configuration then I think there should be documentation here to indicate when this flag needs to be set.

Comment (Alex2782) [reply drivers/vulkan/rendering_device_vulkan.cpp:6089]: `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT`  MacOS would have no problem with this, but I have not checked the performance in detail.

**on Android:**
`Vulkan API 1.1.128 - Forward Mobile - Using Vulkan Device #0: Qualcomm - Adreno (TM) 650`

**on MacOS:**
`Vulkan API 1.2.231 - Forward Mobile - Using Vulkan Device #0: Apple - Apple M1`

------------------------
@sakrel has posted this information, who can test all this?
https://asawicki.info/news_1740_vulkan_memory_types_on_pc_and_how_to_use_them

1. The Intel way
2. The NVIDIA way
3. The AMD way
4. The SAM way
5. The APU way

Comment (Alex2782) [reply drivers/vulkan/rendering_device_vulkan.cpp:6089]: **"Coherent Memory"** could perhaps be useful for testing on more hardware `Rendering -> Rendering Device`? "Staging Buffer" is the other function that also uses `VMA_MEMORY_USAGE_AUTO_PREFER_HOST`. I have not yet checked whether the function can also cause problems on Android.

image

**Comment by RandomShaper** [comment]
I'd suggest simplifyng this PR by making `_buffer_allocate()` always add `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT` in case it's given `VMA_MEMORY_USAGE_AUTO_PREFER_HOST`. Regarding ``VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT`, as far as I know, VMA will add it automatically.

And, as someone else has pointed out, `_insert_staging_block()` would also benefit from `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT` for the CPU-to-GPU direction.

On desktop, you don't need to ask for `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT`, but you should unless you want to rely on driver behavior. Quoting VMA docs:
> Also, Windows drivers from all 3 PC GPU vendors (AMD, Intel, NVIDIA) currently provide `HOST_COHERENT` flag on all memory types that are `HOST_VISIBLE`, so on PC you may not need to bother.

My guess is that there wouldn't be a performance loss on platforms that already add the proper bit. And, on platforms that don't, we have to add it explicitly anyway. Therefore, let's always ask for coherent memory when it comes data transfers via mapped memory.

**Review by clayjohn — DISMISSED** [review]
This approach looks good to me. Let's merge this early in the 4.3 release cycle and cherrypick to 4.2.1

**Comment by Alex2782** [comment]
the new changes have an impact on more functions

`texture_get_data` -> `_buffer_allocate`_(...., VMA_MEMORY_USAGE_AUTO_PREFER_HOST....);_
`_texture_update `and `_buffer_update` -> `_insert_staging_block`

**new Test project** [test.buffer.zip](https://github.com/godotengine/godot/files/13349980/test.buffer.zip)


**MacOs** and Android no problems. 
Godot Android Editor (issue 80371) not tested.

https://github.com/godotengine/godot/assets/41921395/6ba10e6a-0e17-4e39-be63-6558fd558553

**Comment by Alex2782** [comment]
```cpp
#include "thirdparty/vulkan/vk_mem_alloc.h"
```
@akien-mga Can this change be left in?
- not necessary for scons and cmake
- VsCode finds `"vk_mem_alloc.h"`
- Android Studio: `"vk_mem_alloc.h"` (NOT), `<vk_mem_alloc.h>` (NOT)
- `"thirdparty/vulkan/vk_mem_alloc.h"`  (OK), `<thirdparty/vulkan/vk_mem_alloc.h>` (OK)

**Comment by sakrel** [comment]
> The functions `vmaInvalidateAllocation` / `vmaFlushAllocation` require too much performance. The FPS goes from 60 to 40 because 'issue **75599**' queries the data via `_process`. With `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT` this was no problem.
> 
> ```c++
> 	void *buffer_mem;
> 	VkResult vkerr = vmaMapMemory(allocator, tmp_buffer.allocation, &buffer_mem);
> 	ERR_FAIL_COND_V_MSG(vkerr, Vector<uint8_t>(), "vmaMapMemory failed with error " + itos(vkerr) + ".");
> 
> #ifdef __ANDROID__
> 	vmaInvalidateAllocation(allocator, tmp_buffer.allocation, 0, VK_WHOLE_SIZE);
> #endif
> 
> 	Vector<uint8_t> buffer_data;
> 	{
> 		buffer_data.resize(p_size);
> 		uint8_t *w = buffer_data.ptrw();
> 		memcpy(w, buffer_mem, p_size);
> 	}
> 
> #ifdef __ANDROID__
> 	vmaFlushAllocation(allocator, tmp_buffer.allocation, 0, VK_WHOLE_SIZE);
> #endif
> 
> 	vmaUnmapMemory(allocator, tmp_buffer.allocation);
> 
> 	_buffer_free(&tmp_buffer);
> ```
> 
>  Screen_recording_20231113_231241.mp4

hmm, I don't think we should disregard the `vmaInvalidateAllocation` / `HOST_VISIBLE | HOST_CACHED` approach just yet.
The ARM documentation strongly suggests that `HOST_VISIBLE | HOST_CACHED` should be the preferred memory type for CPU readback.
>Use HOST_VISIBLE | HOST_COHERENT | HOST_CACHED memory for resources which are read back on to the CPU. Use HOST_VISIBLE | HOST_CACHED if it is not available.

>Do not read back data from uncached memory on the CPU.

>Uncached readbacks can be much slower than cached reads due to increased CPU processing costs.

>Check that all CPU-read buffers are using cached memory.

https://developer.arm.com/documentation/101897/0301/CPU-overheads/Vulkan-CPU-memory-mapping

However, the documentation also mentions that manually invaliding/flushing the cache can be expensive:
>vkFlushMappedRanges() when the CPU has finished writing data for the GPU.
vkInvalidateMappedRanges() when reading back the data that the GPU has written.
However, both calls are expensive to use, so must be used sparingly.

I wonder if the `vmaInvalidateAllocation` / `HOST_VISIBLE | HOST_CACHED` approach just has a bigger base cost, but scales better with bigger buffer sizes?

Also, I think, the `vmaFlushAllocation` in your code is unnecessary because the CPU never writes to the mapped buffer.

**Comment by clayjohn** [comment]
> hmm, I don't think we should disregard the vmaInvalidateAllocation / HOST_VISIBLE | HOST_CACHED approach just yet.
The ARM documentation strongly suggests that HOST_VISIBLE | HOST_CACHED should be the preferred memory type for CPU readback.

I think you mixed that up. The document you linked recommends using ``HOST_VISIBLE | HOST_COHERENT | HOST_CACHED`` over ``HOST_VISIBLE | HOST_CACHED`` It also recommends avoiding ``vkInvalidateMappedRanges()``

**Comment by clayjohn** [comment]
@darksylinc What is the benefit of using ``HOST_CACHED`` for a one-time only read operation? That's the context we are dealing with here. I agree it makes sense when we will be doing multiple memory accesses before unmapping, but in this specific use-case, is there any point?

**Comment by darksylinc** [comment]
> What is the benefit of using HOST_CACHED for a one-time only read operation?

It's more as to how the CPU operates and _**how**_ we are doing the read.

When we do (in pseudo x86 asm but applies to arm as well):

```asm
loop:
   mov eax, [non_cached_address]
   mov [cached_address], eax
   add non_cached_address, 4
   add cached_address, 4
   goto loop
```

If "non_cached_address" is actually cached, the CPU reads 64 bytes (a cache line) into L1, and probably prefetches much more into L2. So the next 16 iterations (we're reading 4 bytes at a time) read by the mov will be super fast because it's already in L1. On the 17th iteration it will be slightly slower but it still be very fast because it fetches the next 64 bytes from L2. And so on.

If "non_cached_address" is uncached, every 4 bytes is a trip to RAM. I don't know if Out of Order Execution and speculative execution can apply here (if they do, then it can hide several of the next 4 bytes reads). The CPU will have to wait for every read operation to arrive.

Now [memcpy is a tricky one](https://stackoverflow.com/a/33485055), because the insides of it are intrinsic driven based on things like CPU being used on, and size of the transfer in bytes.

But _usually_ memcpy in x86 ends up using either `rep stos` or a loop with a huge vmovaps with AVX2 that reads either 32 or 64 bytes at a time. On aarch64 I suspect there is something similar for bulk reading.

This heavily reduces the impact of a lack of cache because it's bulk-reading from RAM.

**So to summarize:** I suspect that the impact of `HOST_CACHED` (or lack of it) when we just use memcpy is very low.

If Godot had been designed differently, the user would be reading directly from the mapped memory. This would save a lot of bandwidth (because we save a memcpy), but since we don't control how the user going to read that memory, then using  `HOST_CACHED` would be very important.

**Comment by Alex2782** [comment]
+`VK_MEMORY_PROPERTY_HOST_CACHED_BIT`

During my tests, I did not notice any improvements or deterioration.

**Comment by RandomShaper** [comment]
> For writing, you usually don't _need_ `HOST_CACHED` nor `HOST_COHERENT`.
> * Whether you want them depends on how you do the writes.

Don't you need `HOST_COHERENT` for CPU-to-GPU transfer as well (unless you want to explicitly call `vkFlushMappedMemoryRanges()`? Or am I misunderstanding the spec?


Taking the opportunity to ask, is `VK_ACCESS_HOST_WRITE_BIT` needed somewhere for the GPU to read stuff from the staging buffer?

**Comment by RandomShaper** [comment]
> +`VK_MEMORY_PROPERTY_HOST_CACHED_BIT`
> 
> During my tests, I did not notice any improvements or deterioration.


I'd suggest considering the cached flag out of the scope of this PR. This patch is about correctness. We can consider the cached thing a potential optimization belonging to a separate discussion, involving the broader topic of how Godot deals with mapped memory.

**Comment by Alex2782** [comment]
without `HOST_CACHED` again

**Review by clayjohn — APPROVED** [review]
Looks good to me again. 

``HOST_CACHED`` might be useful in the future if we are holding the map to memory on the host side. But right now we just memcpy the memory and move on. ``HOST_VISIBLE | HOST_COHERENT | HOST_CACHED`` is not always available, so it definitely shouldn't be a required memory format. In the future we can add it is a preferred option, if we move away from doing a full memcpy

**Review by RandomShaper — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by YuriSizov** [comment]
Cherry-picked for 4.2.1.
